state/compatibility_test.go

```diff
@@ -98,81 +140,157 @@ func TestSaveLegacyAndLoadFinalizeBlock(t *testing.T) {
-	require.Equal(t, 1, len(loadedABCIResponses.Events))
-	require.Equal(t, len(abciResponses.BeginBlock.Events), len(loadedABCIResponses.Events))
+
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs), len(finalizeBlockResponse.TxResults))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].String(), finalizeBlockResponse.TxResults[0].String())
+	require.Equal(t, len(v1beta2ABCIResponses.BeginBlock.Events), len(finalizeBlockResponse.Events))
 }
 
-// This test unmarshals a v1beta2.ABCIResponses as a statev1.LegacyABCIResponses
-// This logic is important for the LoadFinalizeBlockResponse method in the state sm
+// This test un-marshals a v1beta2.ABCIResponses as a statev1.LegacyABCIResponses
+// This logic is important for the LoadFinalizeBlockResponse method in the state store
 // The conversion should not fail because they are compatible.
-func TestStateProtoV1Beta2ToV1(t *testing.T) {
+//
+//nolint:dupl
+func TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse(t *testing.T) {
 	v1beta2ABCIResponses := newV1Beta2ABCIResponses()
 
 	v1b2Resp, err := v1beta2ABCIResponses.Marshal()
 	require.NoError(t, err)
 	require.NotNil(t, v1b2Resp)
 
 	// un-marshall a v1beta2 ABCI Response as a LegacyABCIResponse
-	legacyABCIResponse := new(statev1.LegacyABCIResponses)
-	err = legacyABCIResponse.Unmarshal(v1b2Resp)
+	legacyABCIResponses := new(statev1.LegacyABCIResponses)
+	err = legacyABCIResponses.Unmarshal(v1b2Resp)
+	require.NoError(t, err)
+
+	// ensure not nil
+	require.NotNil(t, legacyABCIResponses.DeliverTxs)
+	require.NotNil(t, legacyABCIResponses.EndBlock)
+	require.NotNil(t, legacyABCIResponses.BeginBlock)
+
+	// ensure for equality
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs), len(legacyABCIResponses.DeliverTxs))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Code, legacyABCIResponses.DeliverTxs[0].Code)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Data, legacyABCIResponses.DeliverTxs[0].Data)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Log, legacyABCIResponses.DeliverTxs[0].Log)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasWanted, legacyABCIResponses.DeliverTxs[0].GasWanted)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasUsed, legacyABCIResponses.DeliverTxs[0].GasUsed)
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events), len(legacyABCIResponses.DeliverTxs[0].Events))
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes), len(legacyABCIResponses.DeliverTxs[0].Events[0].Attributes))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Key, legacyABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value, legacyABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Codespace, legacyABCIResponses.DeliverTxs[0].Codespace)
```

Comment:
Same comment as above: wouldn't `require.Equal(t, v1beta2ABCIResponses.DeliverTxs, legacyABCIResponses.DeliverTxs)` be checking all elements and all fields one by one?

Reply:
Same answer as above, but here's the error if you try to test for equality

```
=== RUN   TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse
    compatibility_test.go:239: 
        	Error Trace:	/Users/andynogueira/Development/Go/src/github.com/cometbft/cometbft/state/compatibility_test.go:239
        	Error:      	Not equal: 
        	            	expected: []*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x14000255b00)})
        	            	actual  : []*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x14000255b80)})
        	Test:       	TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse
--- FAIL: TestStateV1Beta2ABCIResponsesAsStateV1LegacyABCIResponse (0.00s)

Expected :[]*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x14000255b00)})
Actual   :[]*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x14000255b80)})```

Reply:
Thanks for giving it a try